components/debounce/browser/debounce_throttle.cc

```diff
@@ -0,0 +1,75 @@
+/* Copyright 2021 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/debounce/browser/debounce_throttle.h"
+
+#include <memory>
+
+#include "base/feature_list.h"
+#include "brave/components/brave_shields/browser/brave_shields_util.h"
+#include "brave/components/debounce/browser/debounce_service.h"
+#include "brave/components/debounce/common/features.h"
+#include "components/content_settings/core/browser/host_content_settings_map.h"
+#include "services/network/public/cpp/request_mode.h"
+#include "services/network/public/cpp/resource_request.h"
+#include "url/gurl.h"
+
+namespace debounce {
+
+// static
+std::unique_ptr<DebounceThrottle> DebounceThrottle::MaybeCreateThrottleFor(
+    DebounceService* debounce_service,
+    HostContentSettingsMap* host_content_settings_map) {
+  // If debouncing is disabled in brave://flags, don't create throttle. Caller
+  // must nullcheck this.
+  if (!base::FeatureList::IsEnabled(debounce::features::kBraveDebounce))
+    return nullptr;
+  return std::make_unique<DebounceThrottle>(debounce_service,
+                                            host_content_settings_map);
+}
+
+DebounceThrottle::DebounceThrottle(
+    DebounceService* debounce_service,
+    HostContentSettingsMap* host_content_settings_map)
+    : debounce_service_(debounce_service),
+      host_content_settings_map_(host_content_settings_map) {}
+
+DebounceThrottle::~DebounceThrottle() = default;
+
+void DebounceThrottle::WillStartRequest(network::ResourceRequest* request,
+                                        bool* defer) {
+  if (!brave_shields::ShouldDoDebouncing(host_content_settings_map_,
```

Comment:
@pes10k what site should control whether this is turned on or off? Right now the site that gets skipped is the control and that makes it essentially impossible to disable with shields

Reply:
After discussing on slack, my suggestion is to:
1. not try to tie this to per-site shields settings
2. have a global flag (as already exists)
3. if we later think its important to have per-domain exceptions here, lets do so by adding UI / UX for adding additional global filter list exceptions, and not per-domain content settings changes 
4.

Reply:
@pes10k are we sticking with ^^ or leaving this in place as discussed after realizing this is only supposed to be for top level navigations?

Reply:
for now i think we can just leave as is, and resist or modify shields integration later if needed

Reply:
is it only for top level navigations? I'm probably missing where this is checked in the code